### ISSUE TITLE
Move all documentation to inline

### ISSUE THREAD

**Issue by aendrew** [d3/d3#3052]
So, here's an idea. It's got a bit of a lead in, but please hear me out.

---

*tl;dr* — D3 has no internal documentation comments ("doc-comments"), as far as I can tell from searching through a few modules. This is a tremendous missed opportunity, as I'll detail below. Moving documentation inline would have a bunch of positive synergistic effects, the main two of which are:

### :boom: 1. Improve effectiveness of IDE tooling

Right now, if you use Tern.js, D3's autocompletion looks like this:

<img width="315" alt="screen shot 2017-01-18 at 00 24 35" src="https://cloud.githubusercontent.com/assets/185041/22046006/c256c762-dd15-11e6-86f3-d82d929eaf97.png">

Not bad, pretty useful. But, if you have doc comments next to where you define functions, Tern returns far more information:

<img width="525" alt="screen shot 2017-01-18 at 00 28 29" src="https://cloud.githubusercontent.com/assets/185041/22046013/d93882f4-dd15-11e6-9938-ae1b5ac06553.png">

As per Eric Elliott in [this post](https://medium.com/javascript-scene/you-might-not-need-typescript-or-static-types-aa7cb670a77b), much of TypeScript's ultimate utility is derived from the fact that you can surface API documentation within your IDE, using the TypeScript definitions provided by superheros like Tom Wanzek. But tooling like Tern.js allows a very similar development experience, without the (frankly, as somebody who's done a bit of TypeScript with D3) really annoying extra overhead.

Even if you don't use Tern, it still helps to be able to right-click a module, click "go to source", and then immediately see API documentation.

### :boom: 2. Keeping documentation near the code — and using that as a primary source — will help keep it error-free

Given how the current README/wiki-based documentation is structured, keeping the bits linking to the points in the source code where a function is defined is [quite likely a fairly annoying task](https://github.com/d3/d3/issues/2936). What if there was a programmatic way of generating these links, and even the entirety of the web-based documentation itself?

Inline documentation is one way of doing this. Structured doc comments not only allow for better tooling, it allows for tooling that generates HTML fragments that can be used in web-based documentation systems. This ultimately kills two birds with one stone: as well as keeping references to code points in online documentation up to date, this also helps ensure the documentation stays accurate as it's easier to notice errors when it's right next to the function definition than when it's off sitting in some markdown file somewhere. 

### There's no downside to having documentation in code

We live in the gosh-darn awesome Web 2017 era wherein most JavaScript passes through a bundler and/or transpiler and/or a bunch of other steps before it gets to production, which has the side effect of removing comments in the finished bundle. This means none of the documentation is ever shipped in distributed versions, so D3 bundles remain as lean as ever while adding internal documentation's significant productivity improvements.

### Yes, this is a big project

Moving documentation from the wiki pages can probably be automated in some way, but there will still necessarily have to be quite a lot of review afterwards. A system would then have to be created that populates the README/wiki documentation from the doc comments, so only the doc comments would need to be updated moving forward.

Comparing my productivity in TypeScript with D3 compared to when I'm not using Tern or anything else is fairly telling. At the moment, searching for a function signature as I'm coding generally involves navigating an average of four pages (Google result => D3 wiki "moved" page => D3 v4 overview wiki page => module README). Having documentation inline would dramatically reduce how frequently this needs to be done. For what it's worth, I'm more than happy to contribute to making this happen if it's something people think would be useful. :shipit:

**Comment by ericsoco** [comment]
Documentation (outside of bl.ocks, which are suuuuper helpful) was historically a pain point with d3. V4 docs make a lot of that pain go away... but not when you have to dig into the source. Keeping docs inline yields all the benefits mentioned above, and I'd argue also generally encourages more legible coding. Habitually documenting your code means habitually remembering that people will be reading it.

**Comment by mbostock** [comment]
A small clarification, but there is no longer any API documentation in the wiki; the only official API documentation is in the README.md (or API.md). This change was made with v4 to ensure that the API documentation changes in lockstep with the API, and that the API documentation is versioned. It also allows pull requests to include corresponding documentation changes. So I strongly agree with having the documentation live in the repository, rather than in a separate wiki.

As for putting it directly in the code, I’ve not found a solution I have been happy with. (The only real one I have seen is JavaDoc-style comments.) Yes it was a little tedious to add the source links initially, but it wasn’t overly difficult due to the source code being mostly broken-up into small files. It would be nice to have some validation of the source links, but so far it doesn’t seem like a pain point as the structure doesn’t change very quickly.

My priorities are:

* Implementing new features.
* Fixing bugs.
* Writing more documentation.
* Writing more examples.

I don’t have the bandwidth at the moment to consider changing the documentation format, but you’re welcome to explore that effort and report back what you learn.
